File: scripts/zigux/check-phase4-atomic64-wrapper-evidence.py
#!/usr/bin/env python3
"""Guard Phase 4 atomic64 wrapper rollback evidence.

This checker intentionally stays narrower than the shared Phase 4 gate-evidence
checker.  It verifies that the roadmap-targeted wrapper file remains present,
that the gate-evidence note pins its current Git blob, and that the wrapper still
contains the local source-inventory and shared gate-evidence guard tests that
make rollback reviewable.
"""

from __future__ import annotations

import argparse
import hashlib
import tempfile
from pathlib import Path

NOTE = Path("Documentation/zigux/phase4-gate-evidence.md")
ATOMIC64_DIFF = Path("zigux/tests/atomic64_diff.zig")

PIN_LABEL = "PHASE4_ATOMIC64_DIFF_BLOB_SHA"
REQUIRED_NOTE_MARKERS = (
    "# Phase 4 Gate Evidence",
    "PHASE4_RUNTIME_ATOMIC64_SURVEY_PACKET_PRESENT=true",
    "phase4-runtime-atomic64-diff-survey-tests",
    "make -C zigux phase4-runtime-atomic64-diff-survey",
)
REQUIRED_WRAPPER_MARKERS = (
    "test \"atomic64 diff wrapper keeps the shared gate-evidence packet explicit\" {",
    "test \"atomic64 diff wrapper keeps its own source inventory explicit\" {",
    "PHASE4_ATOMIC64_DIFF_BLOB_SHA={s}",
    "PHASE4_RUNTIME_ATOMIC64_DIFF_BLOB_SHA={s}",
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("root", nargs="?", default=".")
    parser.add_argument("--self-test", action="store_true")
    return parser.parse_args()


def read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def write_text(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def git_blob_sha(path: Path) -> str:
    data = path.read_bytes()
    header = f"blob {len(data)}\0".encode("utf-8")
    return hashlib.sha1(header + data).hexdigest()


def note_values(text: str, label: str) -> list[str]:
    needle = f"`{label}="
    values: list[str] = []
    cursor = 0
    while True:
        start = text.find(needle, cursor)
        if start == -1:
            return values
        value_start = start + len(needle)
        value_end = text.find("`", value_start)
        if value_end == -1:
            return values
        values.append(text[value_start:value_end])
        cursor = value_end + 1


def validate_root(root: Path) -> list[str]:
    failures: list[str] = []
    note_path = root / NOTE
    wrapper_path = root / ATOMIC64_DIFF

    if not note_path.is_file():
        failures.append(f"missing:{NOTE.as_posix()}")
    if not wrapper_path.is_file():
        failures.append(f"missing:{ATOMIC64_DIFF.as_posix()}")
    if failures:
        return failures

    note = read_text(note_path)
    wrapper = read_text(wrapper_path)

    for marker in REQUIRED_NOTE_MARKERS:
        if marker not in note:
            failures.append(f"note:missing:{marker}")
    for marker in REQUIRED_WRAPPER_MARKERS:
        if marker not in wrapper:
            failures.append(f"wrapper:missing:{marker}")

    pins = note_values(note, PIN_LABEL)
    if len(pins) != 1:
        failures.append(f"note:{PIN_LABEL}:count={len(pins)}")
    else:
        expected = git_blob_sha(wrapper_path)
        if pins[0] != expected:
            failures.append(f"note:{PIN_LABEL}:expected={expected}:actual={pins[0]}")

    return failures


def build_fixture(root: Path) -> None:
    wrapper = "\n".join(REQUIRED_WRAPPER_MARKERS) + "\n"
    write_text(root / ATOMIC64_DIFF, wrapper)
    note = "\n".join((
        "# Phase 4 Gate Evidence",
        "",
        "## Status",
        f"  * `{PIN_LABEL}={git_blob_sha(root / ATOMIC64_DIFF)}`",
        "  * `PHASE4_RUNTIME_ATOMIC64_SURVEY_PACKET_PRESENT=true`",
        "",
        "## Exact Readback Evidence",
        "  * The runtime atomic64 handoff remains reviewable through `phase4-runtime-atomic64-diff-survey-tests` and `make -C zigux phase4-runtime-atomic64-diff-survey`.",
        "",
    ))
    write_text(root / NOTE, note)


def run_self_test() -> None:
    cases = 0
    with tempfile.TemporaryDirectory(prefix="phase4-atomic64-wrapper-evidence-") as tmp:
        root = Path(tmp)
        build_fixture(root)
        baseline = validate_root(root)
        if baseline:
            raise AssertionError(f"baseline failed: {baseline}")
        cases += 1

        build_fixture(root)
        (root / ATOMIC64_DIFF).unlink()
        failures = validate_root(root)
        if not any(failure == f"missing:{ATOMIC64_DIFF.as_posix()}" for failure in failures):
            raise AssertionError(f"missing wrapper not detected: {failures}")
        cases += 1

        build_fixture(root)
        write_text(root / NOTE, read_text(root / NOTE).replace(note_values(read_text(root / NOTE), PIN_LABEL)[0], "0" * 40, 1))
        failures = validate_root(root)
        if not any(failure.startswith(f"note:{PIN_LABEL}:expected=") for failure in failures):
            raise AssertionError(f"stale wrapper blob pin not detected: {failures}")
        cases += 1

        build_fixture(root)
        write_text(
            root / ATOMIC64_DIFF,
            read_text(root / ATOMIC64_DIFF).replace(REQUIRED_WRAPPER_MARKERS[0], "", 1),
        )
        failures = validate_root(root)
        if not any(failure.startswith("wrapper:missing:") for failure in failures):
            raise AssertionError(f"wrapper guard marker loss not detected: {failures}")
        cases += 1

    print(f"phase4 atomic64 wrapper evidence self-test: PASS ({cases} cases)")


def main() -> int:
    args = parse_args()
    if args.self_test:
        run_self_test()
        return 0
    failures = validate_root(Path(args.root).resolve())
    if failures:
        for failure in failures:
            print(f"phase4 atomic64 wrapper evidence check failed: {failure}")
        return 1
    print("phase4 atomic64 wrapper evidence check passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase4_atomic64_wrapper_evidence.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const live_pass_marker = "CHECK_PHASE4_ATOMIC64_WRAPPER_EVIDENCE=pass";
pub const self_test_pass_marker = "CHECK_PHASE4_ATOMIC64_WRAPPER_EVIDENCE_SELF_TEST=pass";

const PIN_LABEL = [_][]const u8{
    "PHASE4_ATOMIC64_DIFF_BLOB_SHA",
};

const REQUIRED_NOTE_MARKERS = [_][]const u8{
    "# Phase 4 Gate Evidence",
    "PHASE4_RUNTIME_ATOMIC64_SURVEY_PACKET_PRESENT=true",
    "phase4-runtime-atomic64-diff-survey-tests",
    "make -C zigux phase4-runtime-atomic64-diff-survey",
};

const REQUIRED_WRAPPER_MARKERS = [_][]const u8{
    "test \"atomic64 diff wrapper keeps the shared gate-evidence packet explicit\" {",
    "test \"atomic64 diff wrapper keeps its own source inventory explicit\" {",
    "PHASE4_ATOMIC64_DIFF_BLOB_SHA={s}",
    "PHASE4_RUNTIME_ATOMIC64_DIFF_BLOB_SHA={s}",
};

fn checkRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    const text_pin_label_path = try guard.joinPath(allocator, root, "Documentation/zigux/phase4-gate-evidence.md");
    defer allocator.free(text_pin_label_path);
    const text_pin_label = try guard.readUtf8File(io, allocator, text_pin_label_path);
    defer allocator.free(text_pin_label);
    for (PIN_LABEL) |marker| try guard.requireMarker(text_pin_label, marker);
    const text_required_note_markers_path = try guard.joinPath(allocator, root, "Documentation/zigux/phase4-gate-evidence.md");
    defer allocator.free(text_required_note_markers_path);
    const text_required_note_markers = try guard.readUtf8File(io, allocator, text_required_note_markers_path);
    defer allocator.free(text_required_note_markers);
    for (REQUIRED_NOTE_MARKERS) |marker| try guard.requireMarker(text_required_note_markers, marker);
    const text_required_wrapper_markers_path = try guard.joinPath(allocator, root, "Documentation/zigux/phase4-gate-evidence.md");
    defer allocator.free(text_required_wrapper_markers_path);
    const text_required_wrapper_markers = try guard.readUtf8File(io, allocator, text_required_wrapper_markers_path);
    defer allocator.free(text_required_wrapper_markers);
    for (REQUIRED_WRAPPER_MARKERS) |marker| try guard.requireMarker(text_required_wrapper_markers, marker);
}

fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    try checkRepo(io, allocator, try guard.defaultRepoRoot(allocator));
    try guard.printLine(io, "{s}", .{self_test_pass_marker});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var self_test = false;
    var explicit_root: ?[]const u8 = null;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    const root = explicit_root orelse try guard.repoRootFromScript(allocator);
    defer if (explicit_root == null) allocator.free(root);

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    checkRepo(io, allocator, root) catch {
        std.process.exit(1);
    };
    try guard.printLine(io, "{s}", .{live_pass_marker});
}
